File: rsocket-test/src/main/java/io/rsocket/test/TestSubscriber.java
/*
package io.rsocket.test;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;

import io.rsocket.android.Payload;
import org.mockito.Mockito;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

public class TestSubscriber {
  public static <T> Subscriber<T> create() {
    return create(Long.MAX_VALUE);
  }

  public static <T> Subscriber<T> create(long initialRequest) {
    @SuppressWarnings("unchecked")
    Subscriber<T> mock = mock(Subscriber.class);

    Mockito.doAnswer(
            invocation -> {
              if (initialRequest > 0) {
                ((Subscription) invocation.getArguments()[0]).request(initialRequest);
              }
              return null;
            })
        .when(mock)
        .onSubscribe(any(Subscription.class));

    return mock;
  }

  public static Payload anyPayload() {
    return any(Payload.class);
  }

  public static Subscriber<Payload> createCancelling() {
    @SuppressWarnings("unchecked")
    Subscriber<Payload> mock = mock(Subscriber.class);

    Mockito.doAnswer(
            invocation -> {
              ((Subscription) invocation.getArguments()[0]).cancel();
              return null;
            })
        .when(mock)
        .onSubscribe(any(Subscription.class));

    return mock;
  }
}
*/

File: rsocket-core/src/test/java/io/rsocket/test/util/TestSubscriber.kt
package io.rsocket.test.util

import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.mock

import io.rsocket.Payload
import org.mockito.Mockito
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription

object TestSubscriber {
    @JvmStatic fun <T> create(): Subscriber<T>? = create(java.lang.Long.MAX_VALUE)

    @Suppress("UNCHECKED_CAST")
    fun <T> create(initialRequest: Long): Subscriber<T>? {
        val mock = mock(Subscriber::class.java) as Subscriber<T>

        Mockito.doAnswer { invocation ->
            if (initialRequest > 0) {
                (invocation.arguments[0] as Subscription).request(initialRequest)
            }
            null
        }
                .`when`<Subscriber<T>>(mock)
                .onSubscribe(any(Subscription::class.java))

        return mock
    }

    @JvmStatic fun anyPayload(): Payload {
        return any(Payload::class.java)
    }

    @Suppress("UNCHECKED_CAST")
    fun createCancelling(): Subscriber<Payload>? {
        val mock = mock(Subscriber::class.java) as Subscriber<Payload>

        Mockito.doAnswer { invocation ->
            (invocation.arguments[0] as Subscription).cancel()
            null
        }
                .`when`<Subscriber<Payload>>(mock)
                .onSubscribe(any(Subscription::class.java))

        return mock
    }
}
